usaspending_api/common/etl/spark.py

```diff
@@ -261,3 +261,45 @@ def merge_delta_table(spark: SparkSession, source_df: DataFrame, delta_table_nam
         WHEN NOT MATCHED THEN INSERT *
     """
     )
+
+
+def diff(left: DataFrame, right: DataFrame, unique_key_col="id", compare_cols=[], collect=False):
+    """
+    Compares two dataframes that share a schema and returns difference at a row level.
+    Args:
+        left: assumes a DataFrame wtih a schema that matches the right DataFrame
+        right: assumes a DataFrame wtih a schema that matches the right DataFrame
+        unique_key_col: indidcates a unique indentification column for comparison
+        compare_cols: list of columns to be compared, must exist in both left and right dataframe
+        collect: spark function to limit number of rows returned
+        If compare_cols is left as an empty list, ALL columns will be compared to each other
+        (for rows where unique_key_col values match).
+        Otherwise, specify the columns that determine "sameness" of rows
+    Returns: None
+    """
+
+    if not compare_cols:
+        compare_cols = set(left.schema.names + right.schema.names)
+    distinct_stmts = " ".join([f"WHEN l.{c} IS DISTINCT FROM r.{c} THEN 'C'" for c in compare_cols])
+    compare_expr = f"""
+    CASE
+        WHEN l.exists IS NULL THEN 'I'
+        WHEN r.exists IS NULL THEN 'D'
+        {distinct_stmts}
+        ELSE 'N'
+    END
+    """
+
+    differences = (
+        left.withColumn("exists", lit(1))
+        .alias("l")
+        .join(right.withColumn("exists", lit(1)).alias("r"), left[unique_key_col] == right[unique_key_col], "fullouter")
+        .withColumn("diff", expr(compare_expr))
+    )
+    cols_to_show = ["diff"] + [f"l.{c}" for c in compare_cols] + [f"r.{c}" for c in compare_cols]
+    differences = differences.select(*cols_to_show)
+    # NOTE: .show() shows only a few rows. Call with collect=True to get ALL results, then show
+    if collect:
+        return differences.collect()
+    else:
+        differences.show(truncate=False)
```

Comment:
The return type should be consistent whether `collect` is true/false. In the current version, there is only a return when collect is true. They should either both return or show the differences. Make sure to update the documentation at the top to reflect the change.

Reply:
Updated to clarify

Reply:
The return type was always a bit confused on this. Maybe we just get it merged as is, and then extend/overload it. I'm thinking there should be a base `diff(...) -> DataFrame:` function, and then maybe a `diff_show(...) -> None:` and `diff_collect(...) -> List[Row]:` wrappers that "overload" the base function. 

That would allow folks that want to do further filtering on the differences to get the DataFrame and do so.